File: src/main/java/catwalks/part/PartStair.java
package catwalks.part;

import catwalks.CatwalksMod;
import catwalks.Const;
import catwalks.EnumCatwalkMaterial;
import catwalks.block.EnumCatwalkMaterialOld;
import catwalks.block.EnumDecoration;
import catwalks.part.data.StairSides;
import catwalks.util.meta.*;
import mcmultipart.multipart.Multipart;
import net.minecraft.util.EnumFacing;

/**
 * Created by TheCodeWarrior
 */
public class PartStair extends Multipart implements IDirtyable, IDecoratable {
	public static final String ID = Const.MODID + ":stair";
	
	protected static MetaStorage.Allocator allocator = new MetaStorage.Allocator();
	
	public static ArrayProp<EnumCatwalkMaterial> MATERIAL = CatwalksMod.allocate_material(allocator);
	public static BoolArrayProp DECOR = CatwalksMod.allocate_decor(allocator);
	
	public static ArrayProp<EnumFacing> FACING = allocator.allocateArray("facing", EnumFacing.HORIZONTALS, 2);
	public static BoolMapProp<StairSides> SIDES = allocator.allocateBoolMap("sides", StairSides.values(), StairSides.values().length);
	
	protected MetaStorage storage = new MetaStorage(allocator, this);
	
	public PartStair() {
		super();
	}
	
	//region api stuff
	
	public EnumCatwalkMaterial getCatwalkMaterial() {
		return MATERIAL.get(storage);
	}
	public void setCatwalkMaterial(EnumCatwalkMaterial value) {
		MATERIAL.set(storage, value);
	}
	
	@Override
	public boolean addDecoration(EnumDecoration decor) {
		EnumCatwalkMaterial mat = getCatwalkMaterial();
		int id = mat.getID(decor);
		if(id < 0)
			return false;
		if(DECOR.get(storage, id))
			return false;
		DECOR.set(storage, id, true);
		return true;
	}
	
	@Override
	public boolean removeDecoration(EnumDecoration decor) {
		EnumCatwalkMaterial mat = getCatwalkMaterial();
		int id = mat.getID(decor);
		if(id < 0)
			return false;
		if(DECOR.get(storage, id))
			return false;
		DECOR.set(storage, id, true);
		return true;
	}
	
	@Override
	public boolean hasDecoration(EnumDecoration decor) {
		EnumCatwalkMaterial mat = getCatwalkMaterial();
		int id = mat.getID(decor);
		if(id < 0)
			return false;
		if(DECOR.get(storage, id))
			return false;
		DECOR.set(storage, id, true);
		return true;
	}
	
	public EnumFacing getFacing() {
		return FACING.get(storage);
	}
	public void setFacing(EnumFacing value) {
		FACING.set(storage, value);
	}
	
	public boolean getSide(StairSides side) {
		return SIDES.get(storage, side);
	}
	public void setSide(StairSides side, boolean value) {
		SIDES.set(storage, side, value);
	}
	
	//endregion
	
	@Override
	public void markDirty() {
		super.markDirty();
	}
}

File: src/main/java/catwalks/part/PartStair.kt
package catwalks.part

import catwalks.CatwalksMod
import catwalks.Const
import catwalks.EnumCatwalkMaterial
import catwalks.block.EnumDecoration
import catwalks.part.data.StairSides
import catwalks.util.meta.*
import mcmultipart.multipart.Multipart
import net.minecraft.util.EnumFacing

/**
 * Created by TheCodeWarrior
 */
class PartStair : Multipart(), IDirtyable, IDecoratable {

    protected var storage = MetaStorage(allocator, this)

    //region api stuff

    var catwalkMaterial: EnumCatwalkMaterial
        get() = MATERIAL.get(storage)
        set(value) = MATERIAL.set(storage, value)

    override fun addDecoration(decor: EnumDecoration): Boolean {
        val mat = catwalkMaterial
        val id = mat.getID(decor)
        if (id < 0)
            return false
        if (DECOR.get(storage, id)!!)
            return false
        DECOR.set(storage, id, true)
        return true
    }

    override fun removeDecoration(decor: EnumDecoration): Boolean {
        val mat = catwalkMaterial
        val id = mat.getID(decor)
        if (id < 0)
            return false
        if (DECOR.get(storage, id)!!)
            return false
        DECOR.set(storage, id, true)
        return true
    }

    override fun hasDecoration(decor: EnumDecoration): Boolean {
        val mat = catwalkMaterial
        val id = mat.getID(decor)
        if (id < 0)
            return false
        if (DECOR.get(storage, id)!!)
            return false
        DECOR.set(storage, id, true)
        return true
    }

    var facing: EnumFacing
        get() = FACING.get(storage)
        set(value) = FACING.set(storage, value)

    fun getSide(side: StairSides): Boolean {
        return SIDES.get(storage, side)!!
    }

    fun setSide(side: StairSides, value: Boolean) {
        SIDES.set(storage, side, value)
    }

    //endregion

    override fun markDirty() {
        super.markDirty()
    }

    companion object {
        val ID = Const.MODID + ":stair"

        protected var allocator = MetaStorage.Allocator()

        var MATERIAL = CatwalksMod.allocate_material(allocator)
        var DECOR = CatwalksMod.allocate_decor(allocator)

        var FACING = allocator.allocateArray("facing", EnumFacing.HORIZONTALS, 2)
        var SIDES = allocator.allocateBoolMap("sides", StairSides.values(), StairSides.values().size)
    }
}
